packages/server/src/course/course.controller.ts

```diff
@@ -20,7 +26,51 @@ export class CourseController {
   @Get(':id')
   async get(@Param('id') id: number): Promise<GetCourseResponse> {
     return await CourseModel.findOne(id, {
-      relations: ['officeHours', 'queues'],
+      relations: ['officeHours', 'queues', 'queues.staffList'],
     });
   }
+
+  @Post(':id/ta_location/:room')
+  async checkIn(
+    @Param('id') courseId: number,
+    @Param('room') room: string,
+    @User() user: UserModel,
+  ): Promise<QueuePartial> {
+    // TODO: think of a neat way to make this abstracted
+    const isTAInCourse =
+      (await UserCourseModel.count({
+        where: {
+          role: Role.TA,
+          courseId: courseId,
+          userId: user.id,
+        },
+      })) === 1;
+
+    if (!isTAInCourse) {
+      throw new UnauthorizedException(
+        "Can't check in to office hours for a course you're not a TA of!",
+      );
+    }
+
+    let queue = await QueueModel.findOne(
+      {
+        room,
+        courseId,
+      },
+      { relations: ['staffList'] },
+    );
+
+    if (!queue) {
+      queue = await QueueModel.create({
+        room,
+        courseId,
+        staffList: [],
+      }).save();
+    }
+
```

Comment:
Do we need to check that the user is not already in the `staffList` of the queue? Just incase they hit "Check in" twice?

Reply:
typeorm is smart enough for this :o